### PR TITLE
Animated tiles

### PR DESCRIPTION
**Pull request by 4Denthusiast**
### Contains

Adds animated block tiles, resolving #628.

If a block tile image isn't square, but is the correct shape for a row of squares, it is split into animation frames, which then play automatically when that tile is used (although blocks in the hand or inventory just constantly display the first frame). The block Core:AnimationTest demonstrates this. All of the animation frames are loaded as a continuous sequence in the WorldAtlas, and the vertex shader then switches between them. The only extra data passed to the GPU is the number of animation frames to cycle through for each vertex.

It was discussed in #628 that it may be necessary to restrict the number of frames in the animation to factors of some fixed number. With this solution, that isn't necessary, the only limit on number of frames being space in the world atlas to store them.

A few other things were cleared up along the way in ChunkMesh and ChunkTessellator (which had unclear or misleading comments).

There's a single, small API change in the first commit: BlockMeshPart.mapTexCoords has an additional argument.

The second commit re-implements the lava animations as a standard animated texture, eliminating one of the rendering special cases and removing all references to lava from the engine (except LiquidData, which is unused). As some of the things relating to lava were API, this breaks the API, but in what I think is an obviously desirable way. Unfortunately, the new lava animation is less smooth than the old one, but this is a necessary consequence of rendering it as a normally aligned texture (i.e. pixels that don't move).

### Outstanding before merging
~~The framerate of the block animations is hard-coded in the shader, and relative to world-time. This is consistent with how the lava animation used to work, but seems sometimes undesirable. If the world is slowed down, it makes sense to slow down block animations, but if it's sped up, they may look unpleasant. Also, it would be useful to have this configurable per-block, but I expect the current 6Hz to be suitable for a variety of uses.~~

~~Mostly, the current solution just feels inappropriate.~~
Having block animations change speed with time-dilation is consistent with all the other animations (water waves, waving grass), so I suppose it's probably the best way for it to work.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by Cervator** [comment]
Tested this out. Lava still animates! Yay progress :-)

API hits:

- [ ]  FacadeServer has a hit in `WorldMapResource has a usage where an `int` is now required: `BufferedImage blockImage = Assets.get(blockUrn, BlockTile.class).get().getImage();`
- [ ] DamagingBlocks enables hurtful lava, which relied on the hard coded lava state - that's a bad way to do it and should be updated, it was just the most straight forward way to do it at the time
- [ ] Latest FlowingLiquids still has lava stuff
- [ ] ItemRendering has a usage in `TintOverlayBlockTileProducer with a couple fairly simple looking tweaks needed

I figure three of those are pretty simple, but DamagingBlocks needs a rework, much like Fluid does for the removal of the liquid stuff. I'm not sure exactly how to rework the hurtful lava though, without hard coding it somehow. Would be nice if something similar could be used with Cacti. Both cases could perhaps work with a passive component (non-persistent prefab on lava, cacti, etc) that can be triggered by a "Player is brushing up against a block with this component" type event.

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by 4Denthusiast** [comment]
I've created branches of those modules with the necessary changes, all called "animatedTiles", in my forks (except FlowingLiquids, where it's in the MovingBlocks version, as I have write access to that anyway).

FacadeServer and FlowingLiquids have simple changes. ItemRendering looked like it would have taken a lot of effort to get it to actually use animated tiles, so I just made it ignore the animations. DamagingBlocks had to be mostly re-written, as trying to change only those bits directly affected wouldn't have worked well, and would have been so complicated that there wouldn't be much point in trying to preserve the messy way it used to work (the updated version is also far from ideal, but considerably neater).

File: engine/src/main/java/org/terasology/world/block/tiles/BlockTile.java [hunk 1]
```diff
@@ -37,8 +37,12 @@ public BlockTile(ResourceUrn urn, AssetType<?, TileData> assetType, TileData dat
         reload(data);
     }
 
-    public BufferedImage getImage() {
-        return image;
+    public BufferedImage getImage(int i) {
```
[the remark this answers is not in the document]
Comment (4Denthusiast) [reply engine/src/main/java/org/terasology/world/block/tiles/BlockTile.java]: Good idea. This makes the FacadeServer changes unnecessary.

**Comment by 4Denthusiast** [comment]
I've now made PRs for the module changes, as apparently I forgot to do that before: https://github.com/Terasology/ItemRendering/pull/8, https://github.com/Terasology/DamagingBlocks/pull/9.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by jellysnake** [comment]
All in all looks good to me. I'll merge later today if it doesn't break or get reviewed before then.

**Comment by 4Denthusiast** [comment]
Closes #628

**Pull request merged**